File: node/packages/core/src/sanitize/report.ts
export class SanitizeReport {
  public filename: string;
  public mode: string;
  public author: string | null;

  public tracked_changes_found: number = 0;
  public tracked_changes_accepted: number = 0;
  public tracked_changes_kept: number = 0;
  public change_lines: string[] = [];

  public comments_removed: number = 0;
  public comments_kept: number = 0;
  public removed_comment_lines: string[] = [];
  public kept_comment_lines: string[] = [];

  public metadata_lines: string[] = [];
  public structural_lines: string[] = [];
  public warnings: string[] = [];

  public status: string = "clean";
  public blocked_reason: string | null = null;

  constructor(filename: string, mode: string = "full", author: string | null = null) {
    this.filename = filename;
    this.mode = mode;
    this.author = author;
  }

  public add_transform_lines(lines: string[]) {
    for (const line of lines) {
      const lower = line.toLowerCase();
      if (lower.includes("tracked change") || lower.includes("insertion") || lower.includes("deletion") || lower.includes("accepted")) {
        this.change_lines.push(line);
      } else if (
        lower.includes("author") || lower.includes("template") || lower.includes("company") ||
        lower.includes("manager") || lower.includes("metadata") || lower.includes("timestamp") ||
        lower.includes("custom xml") || lower.includes("custom propert") || lower.includes("identifier") ||
        lower.includes("document variable") ||
        lower.includes("language") || lower.includes("version") ||
        lower.includes("last modified by") || lower.includes("revision count") || lower.includes("last printed") ||
        lower.includes("description/comments")
      ) {
        this.metadata_lines.push(line);
      } else if (lower.includes("comment") || lower.includes("[open]") || lower.includes("[resolved]")) {
        if (lower.includes("kept") || lower.includes("visible")) {
          this.kept_comment_lines.push(line);
        } else {
          this.removed_comment_lines.push(line);
        }
      } else if (lower.includes("hyperlink") || lower.includes("warning")) {
        this.warnings.push(line);
      } else {
        this.structural_lines.push(line);
      }
    }
  }

  public render(): string {
    const sep = "═".repeat(50);
    const lines: string[] = [sep, `Finalization Report: ${this.filename}`];

    const flags: string[] = [];
    if (this.mode === "keep-markup") flags.push("--keep-markup");
    if (this.author) flags.push(`--author "${this.author}"`);
    if (this.tracked_changes_accepted > 0) flags.push("--accept-all");

    if (flags.length > 0) lines.push(flags.join(" "));
    lines.push(sep);

    if (this.status === "blocked") {
      lines.push("");
      lines.push(`BLOCKED: ${this.blocked_reason}`);
      lines.push(sep);
      return lines.join("\n");
    }

    if (this.mode === "keep-markup" && (this.tracked_changes_kept > 0 || this.comments_kept > 0)) {
      lines.push("");
      lines.push("VISIBLE TO COUNTERPARTY");
      if (this.tracked_changes_kept > 0) lines.push(`  Tracked changes: ${this.tracked_changes_kept}`);
      if (this.comments_kept > 0) {
        lines.push(`  Open comments: ${this.comments_kept}`);
        for (const cl of this.kept_comment_lines) lines.push(`    ${cl}`);
      }
      if (this.author) lines.push(`  Author on all markup: "${this.author}"`);
    }

    if (this.change_lines.length > 0) {
      lines.push("");
      lines.push("TRACKED CHANGES");
      for (const cl of this.change_lines) lines.push(`  ${cl}`);
    }

    if (this.removed_comment_lines.length > 0) {
      lines.push("");
      lines.push("COMMENTS (stripped)");
      for (const cl of this.removed_comment_lines) lines.push(`  ${cl}`);
    }

    if (this.metadata_lines.length > 0) {
      lines.push("");
      lines.push("METADATA");
      for (const ml of this.metadata_lines) lines.push(`  ${ml}`);
    }

    if (this.structural_lines.length > 0) {
      lines.push("");
      lines.push("STRUCTURAL & PROTECTION");
      for (const sl of this.structural_lines) lines.push(`  ${sl}`);
    }

    if (this.warnings.length > 0) {
      lines.push("");
      lines.push("WARNINGS");
      for (const w of this.warnings) lines.push(`  ⚠ ${w}`);
    }

    lines.push("");
    lines.push(sep);
    if (this.warnings.length > 0) {
      lines.push(`Result: CLEAN WITH WARNINGS (${this.warnings.length} warning${this.warnings.length > 1 ? 's' : ''})`);
    } else {
      lines.push(`Result: CLEAN (${this.tracked_changes_found} changes resolved, ${this.comments_removed} comments removed)`);
    }
    lines.push(sep);

    return lines.join("\n");
  }
}

File: python/src/adeu/sanitize/report.py
"""
Sanitize report generation.

Produces human-readable reports showing what was stripped, what was kept,
and any warnings — enabling review and sign-off before sending.
"""

from dataclasses import dataclass, field
from typing import Optional


@dataclass
class SanitizeReport:
    """Collects report data during sanitization, then renders as text."""

    filename: str
    mode: str = "full"  # "full", "keep-markup", "baseline"
    author: Optional[str] = None

    # Tracked changes
    tracked_changes_found: int = 0
    tracked_changes_accepted: int = 0
    tracked_changes_kept: int = 0
    change_lines: list[str] = field(default_factory=list)

    # Comments
    comments_removed: int = 0
    comments_kept: int = 0
    removed_comment_lines: list[str] = field(default_factory=list)
    kept_comment_lines: list[str] = field(default_factory=list)

    # Metadata
    metadata_lines: list[str] = field(default_factory=list)

    # Structural
    structural_lines: list[str] = field(default_factory=list)

    # Warnings
    warnings: list[str] = field(default_factory=list)

    # Status
    status: str = "clean"  # "clean", "clean_with_warnings", "blocked"
    blocked_reason: Optional[str] = None

    def add_transform_lines(self, lines: list[str]):
        """Route transform output to the appropriate report section."""
        for line in lines:
            lower = line.lower()
            if any(k in lower for k in ["tracked change", "insertion", "deletion", "accepted"]):
                self.change_lines.append(line)
            elif any(
                k in lower
                for k in [
                    "author",
                    "template",
                    "company",
                    "manager",
                    "metadata",
                    "timestamp",
                    "custom xml",
                    "custom propert",
                    "custom document propert",
                    "document variable",
                    "identifier",
                    "language",
                    "version",
                    "last modified by",
                    "revision count",
                    "last printed",
                    "description/comments",
                ]
            ):
                self.metadata_lines.append(line)
            elif any(k in lower for k in ["comment", "[open]", "[resolved]"]):
                if "kept" in lower or "visible" in lower:
                    self.kept_comment_lines.append(line)
                else:
                    self.removed_comment_lines.append(line)
            elif any(k in lower for k in ["hyperlink", "warning"]):
                self.warnings.append(line)
            else:
                self.structural_lines.append(line)

    def render(self) -> str:
        """Render the full report as text."""
        sep = "═" * 50
        lines = [sep, f"Sanitize Report: {self.filename}"]

        flags = []
        if self.mode == "keep-markup":
            flags.append("--keep-markup")
        elif self.mode == "baseline":
            flags.append("--baseline")
        if self.author:
            flags.append(f'--author "{self.author}"')
        if self.tracked_changes_accepted > 0:
            flags.append("--accept-all")

        if flags:
            lines.append(" ".join(flags))
        lines.append(sep)

        if self.status == "blocked":
            lines.append("")
            lines.append(f"BLOCKED: {self.blocked_reason}")
            lines.append(sep)
            return "\n".join(lines)

        # Visible to counterparty section (for keep-markup / baseline modes)
        if self.mode in ("keep-markup", "baseline") and (self.tracked_changes_kept > 0 or self.comments_kept > 0):
            lines.append("")
            lines.append("VISIBLE TO COUNTERPARTY")
            if self.tracked_changes_kept > 0:
                lines.append(f"  Tracked changes: {self.tracked_changes_kept}")
            if self.comments_kept > 0:
                lines.append(f"  Open comments: {self.comments_kept}")
                for cl in self.kept_comment_lines:
                    lines.append(f"    {cl}")
            if self.author:
                lines.append(f'  Author on all markup: "{self.author}"')

        # Tracked changes section
        if self.change_lines:
            lines.append("")
            lines.append("TRACKED CHANGES")
            for cl in self.change_lines:
                lines.append(f"  {cl}")

        # Stripped section
        stripped_lines = self.removed_comment_lines
        if stripped_lines:
            lines.append("")
            lines.append("COMMENTS (stripped)")
            for cl in stripped_lines:
                lines.append(f"  {cl}")

        # Metadata section
        if self.metadata_lines:
            lines.append("")
            lines.append("METADATA")
            for ml in self.metadata_lines:
                lines.append(f"  {ml}")

        # Structural section
        if self.structural_lines:
            lines.append("")
            lines.append("STRUCTURAL")
            for sl in self.structural_lines:
                lines.append(f"  {sl}")

        # Warnings
        if self.warnings:
            lines.append("")
            lines.append("WARNINGS")
            for w in self.warnings:
                lines.append(f"  ⚠ {w}")

        # Result
        lines.append("")
        lines.append(sep)
        if self.warnings:
            result = f"Result: CLEAN ({len(self.warnings)} warning{'s' if len(self.warnings) > 1 else ''})"
        else:
            result = "Result: CLEAN"
        lines.append(result)
        lines.append(sep)

        return "\n".join(lines)
